source/Irrlicht/CImageLoaderBMP.cpp

```diff
@@ -262,12 +276,16 @@ IImage* CImageLoaderBMP::loadImage(io::IReadFile* file) const
 	// read palette
 
 	long pos = file->getPos();
+	constexpr s32 paletteAllocSize = 256;
 	s32 paletteSize = (header.BitmapDataOffset - pos) / 4;
+	paletteSize = core::clamp(paletteSize, 0, paletteAllocSize);
 
 	s32* paletteData = 0;
 	if (paletteSize)
 	{
-		paletteData = new s32[paletteSize];
+		// always allocate an 8-bit palette to ensure enough space
+		paletteData = new s32[paletteAllocSize];
+		memset(paletteData, 0, paletteAllocSize * sizeof(s32));

```

Comment:
IIRC `new s32[paletteAllocSize]()` or `new s32[paletteAllocSize]{}` would do the zero-initialization as well (unlike bare `new s32[paletteAllocSize]`)